From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 6 Jan 2023 17:37:48 +0900
Subject: [PATCH] Port Very Many Players

Original: ishland <ishlandmc@yeah.net>
Copyright (C) 2023 RelativityMC

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 241bd6cb5f5f37e66a11a4be5c71240eb0f1c72d..5f58832e843f550cb54028851c5b8fe35f1aeda6 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -399,6 +399,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
     public boolean collidingWithWorldBorder; // Paper
     public @Nullable Boolean immuneToFire = null; // Purpur - Fire immune API
+    private boolean boundingBoxChanged = false; // Andromeda - Port VMP
 
     public void setOrigin(@javax.annotation.Nonnull Location location) {
         this.origin = location.toVector();
@@ -1036,6 +1037,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     // Paper end - detailed watchdog information
 
     public void move(MoverType movementType, Vec3 movement) {
+        // Andromeda start - Port VMP
+        if (!boundingBoxChanged && movement.equals(Vec3.ZERO) && this.level.andromedaLevelConfiguration().veryManyPlayers.skipEntityMoveIfNoMovement) {
+            boundingBoxChanged = false;
+            return;
+        }
+        // Andromeda end
         // Paper start - detailed watchdog information
         io.papermc.paper.util.TickThread.ensureTickThread("Cannot move an entity off-main");
         synchronized (this.posLock) {
@@ -3846,6 +3853,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     public final void setBoundingBox(AABB boundingBox) {
+        if (!this.bb.equals(boundingBox) && this.level.andromedaLevelConfiguration().veryManyPlayers.skipEntityMoveIfNoMovement) boundingBoxChanged = true; // Andromeda - Port VMP
         // CraftBukkit start - block invalid bounding boxes
         double minX = boundingBox.minX,
                 minY = boundingBox.minY,
diff --git a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
index 805c816077a591431188d60305a7579f0f04b339..ea64a34897f598823e9535469f7737c1c662955a 100644
--- a/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
+++ b/src/main/java/team/earendel/andromeda/configurations/LevelConfigurations.java
@@ -101,4 +101,11 @@ public class LevelConfigurations extends ConfigurationPart {
 
     }
 
+    public VeryManyPlayers veryManyPlayers;
+    public class VeryManyPlayers extends ConfigurationPart {
+
+        public boolean skipEntityMoveIfNoMovement = false;
+
+    }
+
 }
